ttnn/cpp/ttnn/tensor/tensor_impl.cpp

```diff
@@ -840,6 +847,214 @@ Tensor to_device<bfloat8_b>(
     return to_device<uint32_t>(tensor, target_device, memory_config, cq_id, sub_device_ids);
 }
 
+// ======================================================================================
+//     Helpers for converting between logical <-> physical data with full tensor spec
+// ======================================================================================
+Size flatten_to_2D(const ttnn::SimpleShape& shape) {
+    const int rank = static_cast<int>(shape.rank());
+
+    size_t width = 1;
+    size_t height = 1;
+
+    // Iterate dims in reverse order
+    // Even tensor of rank 0 or 1
+    for (int i = -1; i >= -rank; --i) {
+        auto& dim = i == -1 ? width : height;
+        dim *= shape[i];
+    }
+
+    Size size{height, width};
+
+    return size;
+}
+
+std::array<size_t, 4> compute_shard_spec(const Size& shape, const Size& shard_shape) {
```

Comment:
Can you create a `struct` with named fields instead? (it should work with the structured bindings I see you used below)